tsdb/record/record.go

```diff
@@ -716,10 +744,44 @@ func (e *Encoder) MmapMarkers(markers []RefMmapMarker, b []byte) []byte {
 	return buf.Get()
 }
 
-func (e *Encoder) HistogramSamples(histograms []RefHistogramSample, b []byte) []byte {
+func (e *Encoder) HistogramSamples(histograms []RefHistogramSample, b []byte) ([]byte, []RefHistogramSample) {
 	buf := encoding.Encbuf{B: b}
 	buf.PutByte(byte(HistogramSamples))
 
+	if len(histograms) == 0 {
+		return buf.Get(), nil
+	}
+	var customBucketHistograms []RefHistogramSample
```

Comment:
note: this might be something to optimize away

We could just filter on the fly in this function and also CustomBucketsHistogramSamples. Or we could just overwrite the input and move custom bucket histograms up front in the slice . If I understand correctly it's ok to consume the input slice.

Reply:
Hm I think the main problem with this flow is readability-I had to walk through the code to understand what's happening it's pretty surprising and awkward. However we will fix it in v2 anyway, so I am happy to merge like this -- good enough.

Personally I would prefer your first alternative here, so:

> We could just filter on the fly in this function and also CustomBucketsHistogramSamples.

I think with a clear commentary why ("we will clean in v2")  it would look just more clear. Also I wouldn't filter, but literally skip so `ExponentaitalHistogramSamples` and `CustomBucketsHistogramSamples` functions that just take histograms they are interested in. Not a blocker, up to you what's your preference!

Reply:
One way I have considered making this more clear / not having to return the slice of RefSampleHistograms that are custom bucket histograms from the HistogramSamples() function is to add another 2 fields to the headAppender struct for customBucketsHistograms and customBucketsFloatHistograms (as well as the appenderCommitContext struct). That way, we could add histograms to the histograms or customBucketsHistograms list of RefSampleHistograms at the time of AppendHistogram, and each of the encoding functions would only receive the histograms that correspond to the record type. Any thoughts on this?